### ISSUE TITLE
Window function works incorrect with HistoricalScheduler

### ISSUE THREAD

**Issue by saudar** [dotnet/reactive#870]
Input Data is in 1 min intervals (Dt is interval start):
```
IEnumerable<Item> data = new List<Item> {
                new Item { Id = 1, Value = 1, Dt = DateTime.Parse("2018-12-03 08:00:00")},
                new Item { Id = 2, Value = 1, Dt = DateTime.Parse("2018-12-03 08:01:00")},
                new Item { Id = 3, Value = 1, Dt = DateTime.Parse("2018-12-03 08:02:00")},
                new Item { Id = 4, Value = 1, Dt = DateTime.Parse("2018-12-03 08:03:00")},
                new Item { Id = 5, Value = 1, Dt = DateTime.Parse("2018-12-03 08:04:00")},
                new Item { Id = 6, Value = 1, Dt = DateTime.Parse("2018-12-03 08:05:00")},
                new Item { Id = 7, Value = 1, Dt = DateTime.Parse("2018-12-03 08:06:00")},
       // GAP between 08:07:00 - 08:24:00
                new Item { Id = 8, Value = 1, Dt = DateTime.Parse("2018-12-03 08:25:00")},
                new Item { Id = 9, Value = 1, Dt = DateTime.Parse("2018-12-03 08:26:00")},
                new Item { Id = 10, Value = 1, Dt = DateTime.Parse("2018-12-03 08:27:00")},
                new Item { Id = 11, Value = 1, Dt = DateTime.Parse("2018-12-03 08:28:00")},
                new Item { Id = 12, Value = 1, Dt = DateTime.Parse("2018-12-03 08:29:00")},
                new Item { Id = 13, Value = 1, Dt = DateTime.Parse("2018-12-03 08:30:00")},
            };
```
Historical scheduler is used to replay events with virtual time:
```
var scheduler = new HistoricalScheduler(DateTime.Parse("2018-12-03 08:00:00"));

            var replay = Observable.Generate(
                data.GetEnumerator(),
                events => events.MoveNext(),
                events => events,
                events => events.Current,
                events => events.Current.Dt,
                scheduler);
```
Aggregation to 5 min intervals:
Time is aligned to 5 minutes according 1 hour (0,5,10,15,20,25,30,35,40,45,50,55).
```
            replay
                .Window(TimeSpan.FromMinutes(5), scheduler)
                .SelectMany(stream =>
                    stream
                        .Aggregate(
                            seed: new Item { Id = 0, Value = 0, Dt = scheduler.Now.DateTime },
                            accumulator: (last, newItem) => new Item
                            {
                                Value = last.Value += newItem?.Value ?? 0,
                                Dt = last.Dt
                            }
                        )
                  )
                  .Subscribe(s => Console.WriteLine($"v:{s.Value} t:{s.Dt}"));

            scheduler.Start();
```
But Output is incorrect (v - aggregated value, t - interval start):
```
v:5 t:12/3/2018 8:00:00 AM
v:2 t:12/3/2018 8:05:00 AM
v:0 t:12/3/2018 8:10:00 AM
v:0 t:12/3/2018 8:15:00 AM
v:1 t:12/3/2018 8:20:00 AM
v:4 t:12/3/2018 8:25:00 AM
v:1 t:12/3/2018 8:30:00 AM
```

Intervals: 
```
v:1 t:12/3/2018 8:20:00 AM
v:4 t:12/3/2018 8:25:00 AM
```
are Incorrect, must be:
`v:5 t:12/3/2018 8:25:00 AM`

There are no items in interval, started from 8:20:00, but looks like that, Window function, after gap creates it. And this problem always appears after empty interval.

Is it Bug or special case?

**Comment by bartdesmet** [comment]
Though the behavior may seem unexpected and rather subtle, it is actually by design. The essence is the ordering between two scheduled actions for the same absolute due time:

```csharp
scheduler.Schedule(dueTime, a1);
scheduler.Schedule(dueTime, a2);
```

When `dueTime` is the same, the order of scheduling determines the order of invocation, so `a1` is guaranteed to run before `a2`.

In your example, there are two sources of scheduling:

* `Observable.Generate` used to perform the replay. When a scheduled event `e1` is emitted at time `t1`, the next event `e2` is retrieved from the enumerable source, and scheduled at time `t2`.
* `Window` used to open/close hopping windows. When a window is closed at time `t(i)`, a new window is opened and its closing is scheduled for time `t(i+d)`, where `d` is the window duration.

Now the question boils down to the order in which actions are scheduled by both sources, and in which cases the times coincide exactly.

In your example, fast forward to `2018-12-03 08:06:00`, where an event is emitted by `replay` and the next event to be emitted is retrieved from the enumerable collection, to be emitted at time `2018-12-03 08:25:00`, i.e. after the gap.

Next, fast forward to `2018-12-03 08:20:00`, where a window is closed (having seen `0` events over the last 5 minutes), and a new window is opened with an action scheduled to close it (and open the next window) at `2018-12-03 08:25:00`.

Thus, the order of scheduling is to emit the replayed event first, then to close the window (and open the next window), which explains why the event at time `08:25:00` falls in the preceding window that was opened at `08:20:00`.

The deeper theory of all of this boils down to the fact that Rx does not have special treatment for "punctuation" events (think of these as control plane operations, such as open/close window), e.g. to make these have precedence over data events. Instead, we pick a deterministic FIFO order for any type of event (observed by the scheduler as a `(time, action)` pair) in case of temporal coincidence.

An alternative - which is distinct from Rx semantics - would be to have differentiation of action kinds at the scheduler level and have two buckets of action per time point: "system" events and "data" events. That leads to a distinct set of challenges though, and isn't something we'll consider for Rx given the breaking change nature of it.

In short: if two events coincide at the same time, short of understanding the operational semantics of all operators and the interaction between all sources and operators in a particular query, it's best to think of them as deterministically happening in *either* order and thus not to rely on any particular order.

**Comment by bartdesmet** [comment]
FWIW, here's a type derived from `HistoricalScheduler` that can be used to illustrate the behavior I've described above:

```csharp
class LoggingHistoricalScheduler : HistoricalScheduler
{
    public LoggingHistoricalScheduler(DateTime d)
        : base(d)
    {
    }

    public override IDisposable ScheduleAbsolute<TState>(TState state, DateTimeOffset dueTime, Func<IScheduler, TState, IDisposable> action)
    {
        Console.WriteLine($"{this.Now:HH:mm:ss} - Schedule for {dueTime:HH:mm:ss} - {(state.ToString().Contains("Window") ? "Window" : "Replay")}");
        return base.ScheduleAbsolute(state, dueTime, action);
    }
}
```

When replacing the scheduler by this derived type and running your example, I get the following output:

```
08:00:00 - Schedule for 08:05:00 - Window // Close 08:00-08:05, open 08:05-08:10
08:00:00 - Schedule for 08:00:00 - Replay
08:00:00 - Schedule for 08:00:00 - Replay
08:00:00 - Schedule for 08:01:00 - Replay
08:01:00 - Schedule for 08:02:00 - Replay
08:02:00 - Schedule for 08:03:00 - Replay
08:03:00 - Schedule for 08:04:00 - Replay
08:04:00 - Schedule for 08:05:00 - Replay
v:5 t:12/3/2018 8:00:00 AM
08:05:00 - Schedule for 08:10:00 - Window // Close 08:05-08:10, open 08:10-08:15
08:05:00 - Schedule for 08:06:00 - Replay
08:06:00 - Schedule for 08:25:00 - Replay // Emit event at 08:25 (1)
v:2 t:12/3/2018 8:05:00 AM
08:10:00 - Schedule for 08:15:00 - Window // Close 08:10-08:15, open 08:15-08:20
v:0 t:12/3/2018 8:10:00 AM
08:15:00 - Schedule for 08:20:00 - Window // Close 08:15-08:20, open 08:20-08:25
v:0 t:12/3/2018 8:15:00 AM
08:20:00 - Schedule for 08:25:00 - Window // Close 08:20-08:25, open 08:25-08:30 (2)
08:25:00 - Schedule for 08:26:00 - Replay
v:1 t:12/3/2018 8:20:00 AM
08:25:00 - Schedule for 08:30:00 - Window // Close 08:25-08:30, open 08:30-08:35
08:26:00 - Schedule for 08:27:00 - Replay
08:27:00 - Schedule for 08:28:00 - Replay
08:28:00 - Schedule for 08:29:00 - Replay
08:29:00 - Schedule for 08:30:00 - Replay
v:4 t:12/3/2018 8:25:00 AM
08:30:00 - Schedule for 08:35:00 - Window // Close 08:30-08:35, open 08:35-08:40
v:1 t:12/3/2018 8:30:00 AM
```

Note the order of scheduling of actions `(1)` and `(2)` at `8:25`, which explains the observed behavior.

**Comment by saudar** [comment]
Thanks, for your details explanation - I need more time to analyse and understand it :) 
But I still have question:
if to run this example  with normal scheduler and not virtual, but real time, do i get same results after gap? or it is HistoricalSheduler specific?

**Comment by bartdesmet** [comment]
This behavior should be consistent across well-written schedulers.

First, notice that operators like `Window` do **not** rely on `scheduler.Now` to do their math, e.g. to calculate when the next window needs to be opened or closed. This avoid time drift where the clock is moving forward between the time a scheduled action is invoked and it observes the time again. As such, operators are fully deterministic in the actions they emit on the schedulers they rely on.

Second, local schedulers (those deriving from `LocalScheduler`) use a timer queue that uses the same priority queue implementation used by `HistoricalScheduler`. That's the one that provides behavior to order actions for the same due time in scheduling order.

One of the reasons for introducing `LocalScheduler` is to avoid the implementation of schedulers with strange behavior for absolute due time scheduling, which is fairly complex (also due to possible changes to the system clock). However, someone may have implemented `IScheduler` to just forward a `DateTimeOffset, Action` pair to some underlying facility that does not order actions for the same due time in the way described here.

So, while you can rely on this behavior to be consistent across schedulers provided by Rx, it is best not to rely on anything stronger than the guarantee provided for coincident events, namely that they will be run in any order that is dependent on the composition of operators in the query (as illustrated above for `Window`). Instead, the weaker guarantee you can safely rely on for queries using `Window` is that events that coincide with the closing/opening of a hopping window will be observed by either the previous or the next window, but never both.

Another option is to avoid coincidence of events altogether, which isn't always easy, but sometimes possible. For example, if you know the resolution of your input events is always on millisecond boundaries, you could have your windows align with `0.5` ms boundaries. That obviously only works for historical data or tightly controlled inputs.

Finally, if ambiguity in ordering of coincident events is not tolerable in your system, you could implement a custom scheduler that throws an exception or logs a warning when two events coincide at the same time.

**Comment by bartdesmet** [comment]
Closing this issue given the explanation above on by-design behavior.
